# Сервис для чтения книг

**Цель исследования:**
Проанализировать базу данных. В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.
С помощью SQL-запросов решить задания.

**Задания**

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

**Этапы исследования:**

- Доступ к базе данных;
- Исследование данных;
- Решение задач;
- Выводы.

**Описание данных:**

Таблица *books*

Содержит данные о книгах:

- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

Таблица *authors*

Содержит данные об авторах:

- author_id — идентификатор автора;
- author — имя автора.

Таблица *publishers*

Содержит данные об издательствах:

- publisher_id — идентификатор издательства;
- publisher — название издательства;

Таблица *ratings*

Содержит данные о пользовательских оценках книг:

- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

Таблица *reviews*

Содержит данные о пользовательских обзорах:

- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя автора обзора;
- text — текст обзора.

In [1]:
# Импортируем библиотеки
import pandas as pd
import sqlalchemy as sa

## Доступ к базе данных

In [2]:
# Устанавливаем параметры
db_config = {
'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db' # название базы данных
}
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# Сохраняем коннектор
engine = sa.create_engine(connection_string, connect_args={'sslmode':'require'})

# Чтобы выполнить SQL-запрос, пишем функцию с использованием Pandas
def get_sql_data(query:str, engine:sa.engine.base.Engine=engine) -> pd.DataFrame:
    '''Открываем соединение, получаем данные из sql, закрываем соединение'''
    with engine.connect() as con:
        return pd.read_sql(sql=sa.text(query), con = con)

## Исследование данных

In [9]:
# Посмотрим на данные в таблицах
tables = ['books', 'authors', 'publishers', 'ratings', 'reviews']
for table_name in tables:
    query = f"SELECT * FROM {table_name} LIMIT 5"
    display(get_sql_data(query))

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

Видим что у нас есть пять таблиц с данными о книгах, их авторах, издательствах, проставленными оценками(или рейтингом) и отзывами.

## Решение задач

### Посчитайте, сколько книг вышло после 1 января 2000 года;

In [10]:
# Сформируем запрос и выведе данные
query = '''SELECT count(book_id)
           FROM books
           WHERE publication_date > '2000-01-01' '''
get_sql_data(query)

count
0    819

В результате получаем что после 1 января 2000 года вышла 819 книг.

### Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [11]:
query = '''With amreviews as (SELECT book_id,
                                     count(username) as amount_reviews
                              FROM reviews
                              GROUP BY book_id),
           avrating as (SELECT book_id,
                               ROUND(AVG(rating),2) as avg_ratings
                        FROM ratings
                        GROUP BY book_id)
           SELECT b.title,
                  am.amount_reviews,
                  av.avg_ratings
           FROM books as b
           LEFT JOIN amreviews as am ON b.book_id = am.book_id
           LEFT JOIN avrating as av ON b.book_id = av.book_id'''
get_sql_data(query)

title  amount_reviews  \
0            The Body in the Library (Miss Marple  #3)             2.0   
1                                            Galápagos             2.0   
2                             A Tree Grows in Brooklyn             5.0   
3    Undaunted Courage: The Pioneering First Missio...             2.0   
4                                          The Prophet             4.0   
..                                                 ...             ...   
995                                Alice in Wonderland             4.0   
996          A Woman of Substance (Emma Harte Saga #1)             2.0   
997                                          Christine             3.0   
998  The Magicians' Guild (Black Magician Trilogy  #1)             2.0   
999                           The Plot Against America             2.0   

     avg_ratings  
0           4.50  
1           4.50  
2           4.25  
3           4.00  
4           4.29  
..           ...  
995         4.23  
996         5.00  
997         3.43  
998         3.50  
999         3.00  

[1000 rows x 3 columns]

Для подсчета использовали две отдельные таблицы, после чего с помощью джойна объединили оценки и обзоры со списком книг. И получили таблицу с количеством обзоров и средним рейтингом для всех книг.

In [12]:
query = '''With amreviews as (SELECT book_id,
                                     count(username) as amount_reviews
                              FROM reviews
                              GROUP BY book_id),
           avrating as (SELECT book_id,
                               ROUND(AVG(rating),2) as avg_ratings
                        FROM ratings
                        GROUP BY book_id)
           SELECT b.title,
                  am.amount_reviews,
                  av.avg_ratings
           FROM books as b
           LEFT JOIN amreviews as am ON b.book_id = am.book_id
           LEFT JOIN avrating as av ON b.book_id = av.book_id
           WHERE av.avg_ratings = 5'''
get_sql_data(query)

title  amount_reviews  \
0                                     Light in August             2.0   
1                     Act of Treason (Mitch Rapp  #9)             2.0   
2   The Demon-Haunted World: Science as a Candle i...             2.0   
3                                          Dead Souls             2.0   
4                                               March             2.0   
5   Wherever You Go  There You Are: Mindfulness Me...             2.0   
6                            Neil Gaiman's Neverwhere             2.0   
7                                       Evening Class             2.0   
8   The Ghost Map: The Story of London's Most Terr...             2.0   
9                Welcome to Temptation (Dempseys  #1)             2.0   
10                                         Hard Times             2.0   
11       Moneyball: The Art of Winning an Unfair Game             3.0   
12                  Misty of Chincoteague (Misty  #1)             2.0   
13  The War of Art: Break Through the Blocks & Win...             2.0   
14               Pop Goes the Weasel (Alex Cross  #5)             2.0   
15                 The Big Bad Wolf  (Alex Cross  #9)             2.0   
16                               My Name Is Asher Lev             2.0   
17                          Triptych (Will Trent  #1)             2.0   
18                                 The Woman in Black             2.0   
19  Stone of Farewell (Memory  Sorrow  and Thorn  #2)             1.0   
20      Arrows of the Queen (Heralds of Valdemar  #1)             2.0   
21           Evil Under the Sun (Hercule Poirot  #24)             2.0   
22               Piercing the Darkness (Darkness  #2)             2.0   
23            School's Out—Forever (Maximum Ride  #2)             3.0   
24               Stolen (Women of the Otherworld  #2)             2.0   
25                                          Geek Love             2.0   
26  Captivating: Unveiling the Mystery of a Woman'...             2.0   
27  The Adventures of Tom Sawyer and Adventures of...             1.0   
28  Crucial Conversations: Tools for Talking When ...             2.0   
29  In the Hand of the Goddess (Song of the Liones...             2.0   
30                                Homage to Catalonia             2.0   
31              A Fistful of Charms (The Hollows  #4)             2.0   
32                                        Angels Fall             2.0   
33  How to Be a Domestic Goddess: Baking and the A...             1.0   
34                           Tai-Pan (Asian Saga  #2)             2.0   
35                      A Dirty Job (Grim Reaper  #1)             4.0   
36                       Plum Island (John Corey  #1)             2.0   
37                      Emily of New Moon (Emily  #1)             2.0   
38                                      Alas  Babylon             2.0   
39                                              Women             2.0   
40                                        Marvel 1602             2.0   
41  The Walking Dead  Book One (The Walking Dead #...             2.0   
42          A Woman of Substance (Emma Harte Saga #1)             2.0   
43   The Cat in the Hat and Other Dr. Seuss Favorites             NaN   

    avg_ratings  
0           5.0  
1           5.0  
2           5.0  
3           5.0  
4           5.0  
5           5.0  
6           5.0  
7           5.0  
8           5.0  
9           5.0  
10          5.0  
11          5.0  
12          5.0  
13          5.0  
14          5.0  
15          5.0  
16          5.0  
17          5.0  
18          5.0  
19          5.0  
20          5.0  
21          5.0  
22          5.0  
23          5.0  
24          5.0  
25          5.0  
26          5.0  
27          5.0  
28          5.0  
29          5.0  
30          5.0  
31          5.0  
32          5.0  
33          5.0  
34          5.0  
35          5.0  
36          5.0  
37          5.0  
38          5.0  
39          5.0  
40          5.0  
41    

При этом если выбрать рейтинг равный 5, получим что у нас таких целых 43 книги.

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [13]:
query = '''With publish as (SELECT b.book_id,
                                   p.publisher
                            FROM books as b
                            JOIN publishers as p ON b.publisher_id = p.publisher_id
                            WHERE b.num_pages > 50)
           SELECT publisher,
                  count(book_id) as amount_books
           FROM publish
           GROUP BY publisher
           ORDER BY count(book_id) DESC
           LIMIT 1'''
get_sql_data(query)

publisher  amount_books
0  Penguin Books            42

Для решения задачи объединяем таблицы книг и издательств, сразу убираем книги менее 50 страниц. После чего считаем для каждого издательства количество книг. Сортируем по убыванию и выводим первое издаетльствои из списка. Получаем что искомое издательство - Penguin Books. Издали они 42 книги.

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [14]:
query = '''WITH book_ratings AS (SELECT r.book_id,
                                        AVG(r.rating) AS avg_rating,
                                        COUNT(r.rating) AS total_ratings
                                 FROM ratings r
                                 GROUP BY r.book_id
                                 HAVING COUNT(r.rating) >= 50),
           author_average_ratings AS (SELECT a.author_id,
                                             SUM(br.total_ratings) AS total_ratings,
                                             AVG(br.avg_rating) AS avg_rating
                                      FROM authors a
                                      JOIN books b ON a.author_id = b.author_id
                                      JOIN book_ratings br ON b.book_id = br.book_id
                                      GROUP BY a.author_id)
           SELECT a.author,
                  aar.total_ratings,
                  aar.avg_rating
           FROM authors a
           JOIN author_average_ratings aar ON a.author_id = aar.author_id
           ORDER BY aar.avg_rating DESC
           LIMIT 1'''
get_sql_data(query)

author  total_ratings  avg_rating
0  J.K. Rowling/Mary GrandPré          310.0    4.283844

В результате получили что самую высокую оценку имеет J.K. Rowling/Mary GrandPré. Оценка 4,28, при этом количество оценок 310.

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [15]:
query = '''WITH user_ratings AS (SELECT username,
                                        COUNT(rating_id) AS ratings_count
                                 FROM ratings
                                 GROUP BY username
                                 HAVING COUNT(rating_id) > 48),
           user_reviews AS (SELECT username,
                                   COUNT(review_id) AS reviews_count
                            FROM reviews
                            WHERE username IN (SELECT username FROM user_ratings)
                            GROUP BY username)
           SELECT ROUND(AVG(reviews_count),2) AS average_reviews
           FROM user_reviews'''
get_sql_data(query)

average_reviews
0             24.0

Получаем что среднее количество обзоров пользователей кто поставил более 48 оценок равно 24.

## Выводы

В результе проделанной работы получили следующие выводы:
- После 1 января 2000 года вышло 819 книг. Учитывая тот так что всего книг 1000 можно сказать что это практически 82% от всего списка.
- Для подсчета использовали две отдельные таблицы, после чего с помощью джойна объединили оценки и обзоры со списком книг. И получили таблицу с количеством обзоров и средним рейтингом для всех книг. При этом если выбрать рейтинг равный 5, получим что у нас таких целых 43 книги.
- Наибольшее число книг толще 50 страниц опубликовало издательство Penguin Books. Было издано 42 книги.
- Cамую высокую оценку имеет J.K. Rowling/Mary GrandPré. Оценка 4,28, при этом количество оценок 310.
- А среднее количество обзоров пользователей кто поставил более 48 оценок равно 24. Что говорит нам о том что пользователи кто оставляет оценки так же охотно оставляют обзоры книг.

После получения информации по задачам можно сказать о том, что пользователи продолжают читать. При этом не только оценивать книги, но и оставлять на них свои обзоры, что очень важно и помогает в подборе интересной книги. Считаю что стоит подумать в направлении того чтобы максимально поощрять пользователей делится своим мнение о прочитанных книгах. Так как на освновании имеющихся оценок и обзоров можно собрать рекомендации для других пользователей.